### PR TITLE
Svelte: Improve CSF3 types

### PR DESCRIPTION
**Pull request by kasperpeulen**
## What I did

This PR provides improved type safety for Svelte stories (similar to what we changed to React) but requires svelte-check instead of tsc (Svelte for VS Code extension for editor support):
https://github.com/sveltejs/language-tools/tree/master/packages/svelte-check

Requires the satisfies keyword available in TS ≥ 4.9:
https://github.com/microsoft/TypeScript/pull/46827

Built on top of this CSF PR:
https://github.com/ComponentDriven/csf/pull/51

### Typesafe args
We changed `StoryObj` and `Meta` to increase type safety for when the user provides args partially in meta.

Considering a Component like this:

```svelte
<script lang="ts">
  import { createEventDispatcher } from 'svelte';

  export let disabled: boolean;
  export let label: string;

  const dispatch = createEventDispatcher();
</script>

<button on:click={(e) => dispatch('clicked', e)} on:dblclick on:mousemove {disabled}>
  {label}
</button>
```

It is valid to provide args like this:

```ts
import Button from './Button.svelte';

// ✅ valid
const meta = {
  component: Button,
  args: { disabled: false },
} satisfies Meta<Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Basic = {
  args: { label: 'good' }
} satisfies Story;
```

While it is invalid to forget an arg, in either meta or the story:

```ts
// ❌ invalid
const meta = {
  component: Button,
} satisfies Meta<Button>;

export const Basic = {
  args: { label: 'good' }
} satisfies Story;

// ❌ invalid
const meta = {
  component: Button,
  args: { label: 'good' }
} satisfies Meta<Button>;

export const Basic = {} satisfies Story;
```

Changed Meta to make sure both a Component, as the Props of the component can be used:

```ts
const meta = {
  component: Button,
} satisfies Meta<Button>;

export default meta;

export const meta = {
  component: Button,
} satisfies Meta<ComponentProps<Button>>;

export default meta;
```

If the component is used, you will have enhanced autocompletion for events (maybe in a future version also for slots, when we support that):

```ts
const meta = {
  component: Button,
  args: { label: 'good', disabled: false },
  render: (args) => ({
    Component: Button,
    props: args,
    on: {
      mousemove: (event) => {
        expectTypeOf(event).toEqualTypeOf<CustomEvent>();
      },
    },
  }),
} satisfies Meta<{ disabled: boolean; label: string }>;
```

### Typesafe decorators/loaders

Decorators now accept a new generic arg argument to be specified:

```ts
const withDecorator: DecoratorFn<{ decoratorArg: string }> = (
  storyFn,
  { args: { decoratorArg } }
) => ({
  Component: Decorator,
  props: { decoratorArg },
});
```

And the type of meta/story will check if this arg is part of the generic type:

```ts
type Props = ComponentProps<Button> & { decoratorArg: string };

const meta = satisfies<Meta<Props>>()({
  component: Button,
  args: { disabled: false },
  decorators: [withDecorator],
});

type Story = StoryObj<typeof meta>;
const Basic: Story = { args: { decoratorArg: 0, label: 'good' } };
```

## How to test

You can test it by running: `yarn nx run @storybook/svelte:check`

@benmccann @j3rem1e If you could have a look :)

### CODE REVIEW

File: scripts/utils/options.test.ts [hunk 1]
```diff
@@ -47,6 +48,7 @@ function test(mv: MaybeOptionValues<typeof allOptions>, v: OptionValues<typeof a
 }
 
 describe('getOptions', () => {
+  console.log(maxConcurrentTasks);
```
Comment (benmccann) [inline scripts/utils/options.test.ts:51]: remove this?

**Review by benmccann — COMMENTED** [review]
The PR description mentions TypeScript 4.9, but that's not released yet and it looks like this is using `~4.6.3`?

It looks like there's tons of merge conflicts on this PR

File: scripts/utils/options.test.ts (hunk 1 above)
Comment (kasperpeulen) [reply scripts/utils/options.test.ts]: Thanks, fixed

**Comment by kasperpeulen** [comment]
> The PR description mentions TypeScript 4.9, but that's not released yet and it looks like this is using `~4.6.3`?
> 
> It looks like there's tons of merge conflicts on this PR

@benmccann Yes, I'm working on upgrading to TS 4.9 in this branch:
https://github.com/storybookjs/storybook/pull/19462

But that is blocked by TS 4.9 being released (and the ecosystem around it ESLint and prettier).
I will also update the svelte template in that branch later this week (for `npx sb init`).

This PR can be merged independently of that though, as `satisfies` is only used in user code, not in library code.

Resolving the merge conflicts 😅

File: code/package.json [hunk 2]
```diff
@@ -82,6 +82,7 @@
     "serve-storybooks": "http-server ./built-storybooks -p 8001",
     "smoketest-storybooks": "cross-env STORYBOOK_DISPLAY_WARNING=true DISPLAY_WARNING=true node -r esm ../scripts/smoketest-storybooks.js",
     "storybook:ui": "./lib/cli/bin/index.js dev --port 6006 --config-dir ./ui/.storybook --no-manager-cache",
+    "task": "cd .. && yarn task",
```
Comment (JReinhold) [inline code/package.json:85]: I actually really like this change, but maybe @tmeasday wants to see this.

File: code/package.json [hunk 3]
```diff
@@ -335,7 +336,7 @@
     "ts-jest": "^26.4.4",
     "ts-node": "^10.4.0",
     "tsup": "^6.2.2",
-    "typescript": "4.7.4",
+    "typescript": "~4.6.3",
```
Comment (JReinhold) [inline code/package.json:339]: downgrade?

File: code/renderers/svelte/src/___test___/Button.svelte [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<script lang="ts">
```
Comment (JReinhold) [inline code/renderers/svelte/src/___test___/Button.svelte:1]: These following files have 3 underscores in the directory `/_ _ _test_ _ _/`, is that on purpose? I just did a search in the repo and we don't use that anywhere else.

File: code/renderers/svelte/src/public-types.test.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
+import { describe, test } from '@jest/globals';
```
Comment (JReinhold) [inline code/renderers/svelte/src/public-types.test.ts:1]: I love these type tests. 💪

File: code/renderers/svelte/tsconfig.json [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
   "extends": "../../tsconfig.json",
   "compilerOptions": {
+    "module": "ESNext",
+    "skipLibCheck": true,
     "strict": true,
     "resolveJsonModule": true
   },
   "include": ["src/**/*"],
-  "exclude": ["src/**/*.test.*"]
+  "exclude": []
```
Comment (JReinhold) [inline code/renderers/svelte/tsconfig.json:10]: unless this has an effect, I'd remove this line.

**Review by JReinhold — APPROVED** [review]
This looks really good!

I'm still not sure what the impact is of using `satisfies` and TS 4.9 for our users, but I think you've discussed this earlier with people so I trust that it is under control.

File: code/package.json (hunk 2 above)
Comment (tmeasday) [reply code/package.json:85]: Oh, sure, why not!

File: code/package.json (hunk 3 above)
Comment (ndelangen) [reply code/package.json:339]: More likely ensuring we use the same version everywhere.

File: code/renderers/svelte/tsconfig.json (hunk 6 above)
Comment (ndelangen) [reply code/renderers/svelte/tsconfig.json:10]: I think It's likely required, because of:
https://github.com/storybookjs/storybook/blob/962ed90ccc42509df081a6b969ca6965994b4143/code/tsconfig.json#L20-L29

File: code/package.json (hunk 2 above)
Comment (ndelangen) [inline code/package.json:85]: we should do this in a separate PR

Comment (ndelangen) [reply code/package.json:85]: https://github.com/storybookjs/storybook/pull/19588

Comment (ndelangen) [reply code/package.json:85]: Extracted in a separate PR
